### PR TITLE
MMD-1102 multi staged slim docker builds

### PR DESCRIPTION
**Pull request by cgroeschel**

### CODE REVIEW

**Comment by cgroeschel** [comment]
# to launch the test net 'local' with the default accounts
`cp .env.sample .env
sed -i "s/HAIN_VERSION=dev/HAIN_VERSION=local/" .env.sample
docker-compose -f docker-compose-dev.yml up`

blocks get produced and finalized ok

File: Dockerfile [hunk 1]
```diff
@@ -1,18 +1,13 @@
...
+# create a project folder
 WORKDIR /dhx
 
-# FIXME - only copy necessary files to reduce size of image, and try using intermediate stages again
 COPY . .
 
 RUN apt-get update && apt-get install -y build-essential wget cmake pkg-config libssl-dev \
     openssl git gcc clang libclang-dev curl vim unzip screen bash \
     && curl https://getsubstrate.io -sSf | bash -s -- --fast \
-    && . ~/.cargo/env \
```
Comment (ltfschoen) [inline Dockerfile:15]: i think we need this. we even added it to https://getsubstrate.io/

File: Dockerfile [hunk 2]
```diff
@@ -25,4 +20,11 @@ RUN apt-get update && apt-get install -y build-essential wget cmake pkg-config l
     && rustc --version \
     && cargo build --release
 
+# runtime stage
+FROM rust as runtime
+# set path for docker scripts in case used, to override below default entrypoint
 WORKDIR /dhx/scripts
```
Comment (ltfschoen) [inline Dockerfile:26]:
```suggestion
WORKDIR /dhx/node/scripts
```

File: Dockerfile [hunk 3]
```diff
@@ -25,4 +20,11 @@ RUN apt-get update && apt-get install -y build-essential wget cmake pkg-config l
...
+
+COPY --from=builder /dhx/target/release/datahighway /usr/local/bin
```
Comment (ltfschoen) [inline Dockerfile:28]:
```suggestion
COPY --from=builder /dhx/node/target/release/datahighway /usr/local/bin
```

File: docker-compose-dev.yml [hunk 4]
```diff
@@ -18,6 +18,7 @@ services:
     volumes:
       - basePath:/tmp
       - ./scripts:/dhx/scripts
```
Comment (ltfschoen) [inline docker-compose-dev.yml:20]:
```suggestion
      - ./scripts:/dhx/node/scripts
```

File: docker-compose-dev.yml [hunk 5]
```diff
@@ -33,6 +34,7 @@ services:
...
```
Comment (ltfschoen) [inline docker-compose-dev.yml:36]:
```suggestion
      - ./scripts:/dhx/node/scripts
```

File: docker-compose-dev.yml [hunk 6]
```diff
@@ -48,6 +50,7 @@ services:
...
```
Comment (ltfschoen) [inline docker-compose-dev.yml:52]:
```suggestion
      - ./scripts:/dhx/node/scripts
```

File: docker-compose-dev.yml [hunk 7]
```diff
@@ -78,6 +82,7 @@ services:
...
```
Comment (ltfschoen) [inline docker-compose-dev.yml:84]:
```suggestion
      - ./scripts:/dhx/node/scripts
```

**Review by ltfschoen — CHANGES_REQUESTED** [review]
changes proposed including possible fix to dhx node directory, retain sourcing the .cargo/env file to apply changes after download rust/substrate

File: Dockerfile (hunk 1 above)
Comment (cgroeschel) [reply Dockerfile:15]: this was erroring for me and after removing that line the build went thru fine, so its left out

File: Dockerfile [hunk 8]
```diff
@@ -1,18 +1,13 @@
-FROM debian:buster
-
-ARG CHAIN_VERSION
-ENV _CHAIN_VERSION=${CHAIN_VERSION}
-RUN echo "DataHighway chain version ${_CHAIN_VERSION}"
-
+# build stage
+FROM rust as builder
+# create a project folder
 WORKDIR /dhx
```
[the remark this answers is not in the document]
Comment (cgroeschel) [reply Dockerfile]: updated

File: Dockerfile (hunk 3 above)
Comment (cgroeschel) [reply Dockerfile]: updated

**Review by cgroeschel — COMMENTED** [review]
updated

**Pull request merged**